cargo-ext/src/compile.rs

```diff
@@ -0,0 +1,135 @@
+use anyhow::bail;
+use cargo_metadata::Message;
+use std::path::PathBuf;
+use std::process::Command;
+use std::process::Stdio;
+use std::sync::Arc;
+
+use anyhow::Context;
+use miden_diagnostics::Verbosity;
+use midenc_session::InputFile;
+use midenc_session::OutputFile;
+use midenc_session::OutputType;
+use midenc_session::OutputTypeSpec;
+use midenc_session::OutputTypes;
+use midenc_session::ProjectType;
+use midenc_session::Session;
+use midenc_session::TargetEnv;
+
+pub fn compile(
+    target: TargetEnv,
+    bin_name: Option<String>,
+    output_folder: &PathBuf,
+) -> anyhow::Result<()> {
+    // for cargo env var see https://doc.rust-lang.org/cargo/reference/environment-variables.html
+    let mut cargo_build_cmd = Command::new("cargo");
+    cargo_build_cmd
+        .arg("build")
+        .arg("--release")
+        .arg("--target=wasm32-unknown-unknown");
+    let project_type = if let Some(ref bin_name) = bin_name {
+        cargo_build_cmd.arg("--bin").arg(bin_name.clone());
```

Comment:
NOTE: My later comment that goes into how we determine what target to build and all that using Cargo metadata covers this as well, but my point about following the conventions of `cargo build` still holds in terms of choosing defaults IMO.

Reply:
> If we want to allow specifying a target, we should follow the conventions of `cargo build` I think, to avoid having different defaults than expected.

I like the idea of mirroring `cargo build` options and semantics! It got me thinking about taking the approach of `cargo component` for building the Wasm component - just pass all the options to `cargo` and pick up Wasm files afterward. See  https://github.com/bytecodealliance/cargo-component/blob/87949555d067e92a1872c03daa272794a6c0f6a5/src/lib.rs#L132-L180
I think we can take the same approach for our `build`(`compile`) command. This means for our `build` command, we'll have only `--miden-target` option for VM vs. rollup (I'm not even sure we need that), and the rest will be passed to `cargo build` as is adding `--target wasm32-wasi` if it's not present. I believe the Wasm component binary alone would suffice to figure out if Miden program or library is being built.

Reply:
I think I'm fine with following `cargo component` here, though I think it remains to be seen whether our more complex build process forces us to be more restrictive than they are. That said, it feels like a better starting point for sure, and will probably get us quite far before we need to do anything special.